### DISCUSSION TITLE
Support for Custom Response Header

### DISCUSSION THREAD

**Discussion by brandonsmith86** [dotnet/aspnet-api-versioning#895] [Q&A] [answered]
Our organization has a requirement to include the planned retirement date for a deprecated version.  This information would be included in the response header, much like the deprecated-versions and supported-versions headers are now.  I was hoping to simply extend the existing functionality, but I am not seeing any documentation, or even examples, of how to do that.

Can anyone tell me if this is possible, or do I need to roll my own so to speak?

**Comment by commonsensesoftware** [comment, marked as answer]
Starting in `6.0` this is now possible. The feature _Sunset Policies_ implements [RFC 8594](https://datatracker.ietf.org/doc/html/rfc8594), which satisfies this requirement. [RFC 8288](https://datatracker.ietf.org/doc/html/rfc8288) is also supported so that you can provide additional links to published policies. It's worth noting that an API doesn't have to be deprecated to advertise a sunset date. You might know ahead of time that a supported API will deprecate and then fully sunset well in advance (ex: a year from now).

I'm behind on updating the documentation to fully outline the feature, but there is an open issue tracking that. You can see an example of how to use it [here](../blob/main/examples/AspNetCore/WebApi/OpenApiExample/Program.cs#L20) or take a look at the tests.

There is also now a client library to complement `HttpClient` which can detect this information and log information and warnings when new versions become available, versions are deprecated, or versions are about to sunset.

Happy to answer any more questions about how the feature works. In general, you just configure it from the `ApiVersioningOptions`. The policy can target a specific API, version, or combination of both. You still have to enable `ReportApiVersions` to output this information. I thought about separating it out, but I can't imagine why you'd want sunset policies separate from supported and deprecated versions.

**Reply by brandonsmith86** [reply]
This is great info.  I will dig in tomorrow and get back to you with any questions.

**Comment by brandonsmith86** [comment]
After switching to the new version (v6.1), I am getting a Swashbuckle error: `Conflicting method/path combination "GET DownstreamPath/{node_id}" for actions - neo4j_api.Controllers.QueryController.DownstreamPath (neo4j-api), neo4j_api.Controllers.QueryController.NewDownstreamPath`.  Seems it can no longer differentiate between different versions of the same endpoint.

Here's what my controller code looks like.  The new version basically does nothing.  This is simply to test the versioning functionaity.
``` cs
[ApiController]
[Route("/")]
[ApiVersion(ApiVersions.V1_0, Deprecated = ApiVersions.V1_0_Deprecated)]
[ApiVersion(ApiVersions.V1_1, Deprecated = ApiVersions.V1_1_Deprecated)]
public class QueryController : ControllerBase
{
	/// <summary>
	/// GET request that returns a path of nodes downstream starting from the node specified as a parameter
	/// </summary>
	/// <returns>
	/// A path type which is derived from a list of Nodes and Relationships
	/// </returns>
	/// <param name="node_id">Name of starting node</param>
	/// <param name="limit_hops">Max number of hops to traverse from <paramref name="node_id"/></param>
	/// <param name="limit_nodes">Max number of paths to traverse</param>
	[Route("/DownstreamPath/{node_id}")]
	[HttpGet]		
	[MapToApiVersion(ApiVersions.V1_0)]
	[EnableQuery(AllowedQueryOptions = AllowedQueryOptions.Filter | AllowedQueryOptions.Select | AllowedQueryOptions.Expand)]
	public async Task<Models.v1_0.ApiResponse<Edge>> DownstreamPath(string node_id, int limit_hops = 2, int limit_nodes = 100)
	{
		// v1.0 code here
	}

	/// <summary>
	/// Version 1.1 of GET request that returns a path of downstream nodes
	/// </summary>
	/// <returns>
	///		An error message saying this version hasn't yet been implemented
	/// </returns>
	[Route("/DownstreamPath/{node_id}")]
	[HttpGet]
	[MapToApiVersion(ApiVersions.V1_1)]
	[EnableQuery(AllowedQueryOptions = AllowedQueryOptions.Filter | AllowedQueryOptions.Select | AllowedQueryOptions.Expand)]
	public Task<Models.v1_1.ApiResponse<Edge>> NewDownstreamPath(string node_id, int limit_hops = 2, int limit_nodes = 100)
	{
		// v1.1 code here
	}
}
```

**Reply by brandonsmith86** [reply]
@commonsensesoftware Just tagging you to make sure you see this.

**Reply by brandonsmith86** [reply]
I also tried separating the versions out into separate controllers, but I am still getting the same error.

**Comment by xavierjohn** [comment]
Here is a working example of multiple versions on the same controller.
https://github.com/xavierjohn/ApiVersionSample/blob/main/OpenApiExample/src/V1/Controllers/OrdersController.cs

hosted here
https://openapiversionsample.azurewebsites.net

**Reply by brandonsmith86** [reply]
This uses the old version.   I have no problems with my code when I revert back to the old version.  This only happens on the new version (v6.1).

**Reply by xavierjohn** [reply]
I am using the latest.

< PackageReference Include="Asp.Versioning.Mvc.ApiExplorer" Version="6.1.0" / >

**Reply by brandonsmith86** [reply]
I just confirmed that your sample works on my machine.  There is apparently something going on with my code that is causing this.   In my `ConfigureSwaggerOptions` class, the version provider is only picking up one of the versions, and I'm not sure why.  Very strange.  As I mentioned, it worked fine on the older version.

**Reply by commonsensesoftware** [reply]
The new configuration introduces `IApiVersioningBuilder` which connects all of the related to pieces together. That wasn't so obvious in the past. Several people have missed or been confused by `.AddMvc`. That doesn't add full MVC, just MVC Core. This is new and necessary to support _Minimal APIs_ versus MVC Core (with controllers). In previous versions, that wasn't necessary.

The setup should be as follows:

>[Revised] I notice you are using OData as well

```c#
services.AddApiVersioning()     // Core API Versioning services with support for Minimal APIs
        .AddMvc()               // API version-aware extensions for MVC Core
        .AddODataApiExplorer(); // API version-aware API Explorer extensions for OData
```

The same or similar names to `IMvcCoreBuilder`, but completely different behavior.

Since you are only using _some_ OData (e.g. query options), you may want to review the [Some OData OpenAPI Example](../blob/main/examples/AspNetCore/OData/SomeODataOpenApiExample).

**Reply by brandonsmith86** [reply]
Thank you @commonsensesoftware.  That was helpful.  I got past that error and am now getting the below.  I am digging in to the example you gave as well.
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Asp.Versioning.ApiExplorer.ODataApiDescriptionProvider.TryMatchModelVersion(ApiDescription description, IReadOnlyList`1 items, IODataRoutingMetadata& metadata)
   at Asp.Versioning.ApiExplorer.ODataApiDescriptionProvider.OnProvidersExecuted(ApiDescriptionProviderContext context)
   at Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescriptionGroupCollectionProvider.GetCollection(ActionDescriptorCollection actionDescriptors)
   at Microsoft.AspNetCore.Mvc.ApiExplorer.ApiDescriptionGroupCollectionProvider.get_ApiDescriptionGroups()
   at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GetSwaggerDocumentWithoutFilters(String documentName, String host, String basePath)
   at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GetSwaggerAsync(String documentName, String host, String basePath)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Watch.BrowserRefresh.BrowserRefreshMiddleware.InvokeAsync(HttpContext context)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()
```

**Reply by commonsensesoftware** [reply]
Great!  This looks like it is due to #891, which is a bug. I have the fix, but it isn't published yet. The following should work around the problem:

```c#
services.AddControllers()
        .AddOData(
            options =>
            {
                options.RouteOptions.EnableKeyInParenthesis = false;
                options.RouteOptions.EnableKeyAsSegment = true;
            });
```

**Reply by brandonsmith86** [reply]
This did not resolve the problem.  Still getting the same error.  I really appreciate you guys spending the time to help me with this!

**Reply by brandonsmith86** [reply]
It's working!  I commented out the below line and Swagger started working again.  I'm assuming there's something I'm missing from your OData example?

``` cs 
services.AddControllers().AddOData(options => {
  options.Select().Filter().Expand(); //.AddRouteComponents("odata", EdmModelBuilder.Build());
  options.RouteOptions.EnableKeyInParenthesis = false;
  options.RouteOptions.EnableKeyAsSegment = true;
})
```

**Reply by commonsensesoftware** [reply]
Nice!

This happens because the native `.AddRouteComponents` will not work with API Versioning. Despite previous discussions with the OData team, they elected not to make their design extensible. It expects one EDM per route. API Versioning, however, allows one EDM per API version per route. Nearly every major version change of OData has required some level of refactoring for API Versioning support. 8.0 was a completely rewrite, which is a big part of why it took so long to release.

The API Versioning setup for OData is intentionally very similar:

```c#
// standard OData setup
// the only things honored here are global settings (so you don't have to do it multiple times)
services.AddControllers().AddOData();

// versioned OData setup
// EDMs are constructed from IModelConfiguration instances resolved from DI or
// you can use ODataApiVersioningOptions.ModelBuilder
services.AddApiVersioning().AddOData(options => options.AddRouteComponents("odata"));
```

**Reply by brandonsmith86** [reply]
That works.  Thank you!  No further questions!
